## Avaliação Continuada (AC2) - Perceptron

Nesse notebook, iremos ver como aplicar o método Perceptron na nossa base de dados que foi processada no notebook `train_data.ipynb`.

Para começar então, vamos importar o notebook `train_data.ipynb` que irá conter tudo que precisamos.

In [1]:
import import_ipynb
import train_data

importing Jupyter notebook from train_data.ipynb


Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

Category                 Message
count      5572                    5572
unique        2                    5157
top         ham  Sorry, I'll call later
freq       4825                      30

Shape antes de remover os dados.
(5572, 2)

Shape após remover os dados.
(5157, 2)

Quantidade de itens removidos: 415
Um dos textos antes do processamento:


'I sent you  &lt;#&gt;  bucks'

O mesmo texto, só que agora processado:


'sent you bucks'

Algumas palavras que iremos usar para montar o nosso vocabulário:
['go', 'until', 'jurong', 'point', 'crazy', 'available', 'only', 'in', 'bugis', 'great', 'world', 'la', 'buffet', 'cine', 'there', 'got', 'amore', 'wat']
Tamanho do Vocabulario:  3632
[ 0.17298059 -0.12730037 -0.08990677  0.1658044   0.2830998  -0.35171896
  0.20532139  0.6496812  -0.19397533  0.23911075 -0.24672031 -0.39574838
  0.12225588  0.20485836 -0.09067287 -0.35339254 -0.08169971 -0.02069619
  0.02989911 -0.5034012   0.23397681 -0.31133887 -0.16868372 -0.1244454
  0.09479164 -0.24028893 -0.04909417 -0.1424595  -0.3441023   0.14649226
  0.40862703  0.01599889  0.166184   -0.08662187 -0.07782395  0.2334802
  0.28667802 -0.13569346 -0.07085948 -0.5217293  -0.33183533 -0.09353404
  0.00412842  0.11887     0.5155706  -0.14765029 -0.22448291 -0.09404226
  0.16624415  0.21626587  0.01244251 -0.15770991 -0.22843757 -0.29258433
  0.08290733 -0.09509762 -0.00700543 -0.31515178 -0.32085517  0.1416506
 -0.00293808  0.0473338

## Treinamento

Agora, com os dados separados, vamos começar o treinamento usando o método Perceptron.

Primeiro, importamos o método que iremos usar.

In [2]:
import sklearn
from sklearn.linear_model import Perceptron

Depois, podemos definir algumas variáveis para customizar o comportamento do nosso modelo.

In [3]:
n_interations = 100 # vai controlar a quantidade de interações que será realizada no modelo, também chamado de epochs
learning_rate = 0.1 # a constante que representa a taxa de aprendizado

Depois, criamos o nosso modelo da seguinte forma:

In [4]:
perceptron = Perceptron(max_iter=n_interations, eta0=learning_rate, random_state=0)

E ai, já podemos começar o treinamento do modelo.

In [5]:
perceptron.fit(train_data.x_train, train_data.y_train) 

Perceptron(eta0=0.1, max_iter=100)

E com o modelo treinado, podemos começar a ver as métricas de acurácia do nosso modelo tentando predizer os valores de teste que separamos na etapa anterior.

In [6]:
y_pred = perceptron.predict(train_data.x_test) 

Com os valores que foram classificados acima, podemos ver qual foi a sua performance usando o `sklearn` da seguinte forma:

In [7]:
results = sklearn.metrics.classification_report(train_data.y_test, y_pred, target_names=['ham', 'spam'])

print(results)

              precision    recall  f1-score   support

         ham       0.96      0.95      0.96      1352
        spam       0.69      0.76      0.72       196

    accuracy                           0.93      1548
   macro avg       0.83      0.86      0.84      1548
weighted avg       0.93      0.93      0.93      1548



Com a tabela acima, podemos tirar algumas conclusões como: 

A precisão para identificar mensagens verdadeiras é maior do que a precisão do que identificar mensagens de spam. E isso não necessariamente é ruim porque é melhor errar mais mensagens de spam do que mensagens verdadeiras.

Dessa forma, no pior dos casos ainda vemos o spam e ignoramos mas com a mensagem real indo parar no spam, é mais dificil de ir lá olhar.

E além disso, podemos checar a nossa acurácia com mais precisão tanto dos valores de teste quanto dos valores de treino da seguinte forma:

In [8]:
accuracyTest = train_data.accuracy_score(train_data.y_test, perceptron.predict(train_data.x_test))
accuracyTrain = train_data.accuracy_score(train_data.y_train, perceptron.predict(train_data.x_train))

print("Acurácia sobre os valores de teste: ", accuracyTest)
print("Acurácia sobre os valores de treino: ", accuracyTrain)

Acurácia sobre os valores de teste:  0.9263565891472868
Acurácia sobre os valores de treino:  0.9348850096979773


E aqui vemos, temos, assim como mostrado no relatório anterior, certa de 93% e 93% de acurácia sobre os valores de teste e de treino, respectivamente.

## Cross-Validation

Apesar de eu não entender tão bem como funciona a validação cruzada, eu quis trazer nesses métodos para ter um pequena noção de como a ordem dos dados de treinamento influenciam o nosso modelo.

Dessa forma, o cross-validation permite que a gente separe em partições os nossos dados, e eles por consequencia, serão usados para treinar várias vezes o modelo e assim podemos obter as métrica de acurácia de cada vez que foi treinado o modelo.

Com essas métricas, podemos verificar se a mudança na ordem dos dados tem uma mudança significativa na acurácia do nosso modelo.

Vamos começar então com a importação das bibliotecas que iremos usar.

In [9]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit

Depois, vamos criar algumas variáveis para configurarmos como será a cross-validation.

In [10]:
n_splits = 5 # o número de partições
test_size = 0.3 # a porcentagem que será usada para teste, o restante é usada no treino
cv = ShuffleSplit(n_splits=n_splits, test_size=test_size, random_state=0) # define a função que irá realizar o rearranjo dos dados de treino e como será divido os dados.

E assim, podemos agora chamar a função `cross_val_score` que irá nos retornar a acurácia, assim como, o desvio da acurácia que nos diz quanto a ordem os dados está influenciando o modelo.

In [11]:
scores = cross_val_score(perceptron, train_data.x, train_data.y, cv=cv)

print("%0.2f acurácia com o desvio de %0.2f" % (scores.mean(), scores.std()))

0.84 acurácia com o desvio de 0.13


É interessante ver que agora a acurácia do modelo saiu de 93% para 84%, com um erro de 0.02. 

Isso foi uma piora estranha comparada a outras execuções, eu executei isso em dois dias diferentes e tive resultados diferentes, apesar de setar e tentar deixar os mesmos estados randomicos (random_state=0), ele ainda teve uma diferença ao ser executado em dias diferentes.

E professor, se você chegou até aqui, seria interessante ver mais sobre cross-validation nas aulas.